shell/platform/darwin/graphics/FlutterDarwinContextMetal.mm

```diff
@@ -0,0 +1,67 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/graphics/FlutterDarwinContextMetal.h"
+
+#include "flutter/common/graphics/persistent_cache.h"
+#include "flutter/fml/logging.h"
+#include "third_party/skia/include/gpu/GrContextOptions.h"
+
+static GrContextOptions CreateMetalGrContextOptions() {
+  GrContextOptions options = {};
+  if (flutter::PersistentCache::cache_sksl()) {
+    options.fShaderCacheStrategy = GrContextOptions::ShaderCacheStrategy::kSkSL;
+  }
+  flutter::PersistentCache::MarkStrategySet();
+  options.fPersistentCache = flutter::PersistentCache::GetCacheForProcess();
+  return options;
+}
+
+@implementation FlutterDarwinContextMetal
+
+- (instancetype)initWithDefaultMTLDevice {
+  id<MTLDevice> mtlDevice = MTLCreateSystemDefaultDevice();
+  return [self initWithMTLDevice:mtlDevice commandQueue:[mtlDevice newCommandQueue]];
+}
+
+- (instancetype)initWithMTLDevice:(id<MTLDevice>)mtlDevice
+                     commandQueue:(id<MTLCommandQueue>)commandQueue {
+  self = [super init];
+  if (self != nil) {
+    _mtlDevice = mtlDevice;
+
+    if (!_mtlDevice) {
+      FML_DLOG(ERROR) << "Could not acquire Metal device.";
+      return nil;
+    }
+
+    _mtlCommandQueue = commandQueue;
+
+    if (!_mtlCommandQueue) {
+      FML_DLOG(ERROR) << "Could not create Metal command queue.";
+      return nil;
+    }
+
+    [_mtlCommandQueue setLabel:@"Flutter Main Queue"];
+
+    auto contextOptions = CreateMetalGrContextOptions();
+
+    // Skia expect arguments to `MakeMetal` transfer ownership of the reference in for release later
+    // when the GrDirectContext is collected.
+    _mainContext =
```

Comment:
Does the explicitly retained device and command queue need to be released in case of the nil return later? I am not sure if the Skia API releases them for us in case of error.

Reply:
Oh, you just moved this bit of code. Nevermind. I guess just the releases in case of early returns need to be added in this version because the earlier function was a C++ ctor instead of an ObjC one here.

Reply:
Done